### PR TITLE
[health_check] Pass the result of the current health check run to the hc callbacks

### LINKED ISSUE #32442: successful_active_health_check_uneject_host can cause not wanted outlier uneject
Title: `successful_active_health_check_uneject_host` can cause not wanted outlier uneject

Description:
This callback function:
https://github.com/envoyproxy/envoy/blob/b5ca88acee3453c9459474b8f22215796eff4dde/source/common/upstream/outlier_detection_impl.cc#L315
doesn't care about the result of the health check or the health transition parameter, only the health flags.

Consider the following scenario:
1. The peer host becomes not available.
2. `FAILED_ACTIVE_HC` health flag is still not set because there was no or not enough failed health check so far.
3. Outlier detection ejects the host.
4. If the next health check still doesn't set `FAILED_ACTIVE_HC`, it will uneject the host regardless of whether the check result was success or failure.

You can see this in a real health check and outlier detection log:

Host `127.0.0.1:40002` is down, first failed health check:
```
{"health_checker_type":"THRIFT","host":{"socket_address":{"protocol":"TCP","address":"127.0.0.1","port_value":40002,"resolver_name":"","ipv4_compat":false}},"cluster_name":"cluster_0","timestamp":"2024-02-13T16:15:01.267Z","health_check_failure_event":{"failure_type":"NETWORK","first_check":false},"locality":{"region":"","zone":"","sub_zone":""}}
```

Outlier detection ejects the host:
```
{"type":"CONSECUTIVE_5XX","timestamp":"2024-02-13T16:15:03.083Z","cluster_name":"cluster_0","upstream_url":"127.0.0.1:40002","action":"EJECT","num_ejections":1,"enforced":true,"eject_consecutive_event":{}}
```

Second failed health check. Callback unejects the host which makes its state healthy, but this is obviousy wrong:
```
{"health_checker_type":"THRIFT","host":{"socket_address":{"protocol":"TCP","address":"127.0.0.1","port_value":40002,"resolver_name":"","ipv4_compat":false}},"cluster_name":"cluster_0","timestamp":"2024-02-13T16:15:06.268Z","health_check_failure_event":{"failure_type":"NETWORK","first_check":false},"locality":{"region":"","zone":"","sub_zone":""}}
{"type":"CONSECUTIVE_5XX","timestamp":"2024-02-13T16:15:06.268Z","secs_since_last_action":"3","cluster_name":"cluster_0","upstream_url":"127.0.0.1:40002","action":"UNEJECT","num_ejections":1,"enforced":false}
```

Outlier detectiont ejects the host again:
```
{"type":"CONSECUTIVE_5XX","timestamp":"2024-02-13T16:15:09.084Z","secs_since_last_action":"2","cluster_name":"cluster_0","upstream_url":"127.0.0.1:40002","action":"EJECT","num_ejections":2,"enforced":true,"eject_consecutive_event":{}}
```

Third failed health check finally sets `FAILED_ACTIVE_HC`, callback does nothing just like at the further health checks.
```
{"health_checker_type":"THRIFT","host":{"socket_address":{"protocol":"TCP","address":"127.0.0.1","port_value":40002,"resolver_name":"","ipv4_compat":false}},"cluster_name":"cluster_0","eject_unhealthy_event":{"failure_type":"NETWORK"},"timestamp":"2024-02-13T16:15:11.269Z","locality":{"region":"","zone":"","sub_zone":""}}
{"health_checker_type":"THRIFT","host":{"socket_address":{"protocol":"TCP","address":"127.0.0.1","port_value":40002,"resolver_name":"","ipv4_compat":false}},"cluster_name":"cluster_0","timestamp":"2024-02-13T16:15:11.269Z","health_check_failure_event":{"failure_type":"NETWORK","first_check":false},"locality":{"region":"","zone":"","sub_zone":""}}
{"health_checker_type":"THRIFT","host":{"socket_address":{"protocol":"TCP","address":"127.0.0.1","port_value":40002,"resolver_name":"","ipv4_compat":false}},"cluster_name":"cluster_0","timestamp":"2024-02-13T16:15:16.270Z","health_check_failure_event":{"failure_type":"NETWORK","first_check":false},"locality":{"region":"","zone":"","sub_zone":""}}
{"health_checker_type":"THRIFT","host":{"socket_address":{"protocol":"TCP","address":"127.0.0.1","port_value":40002,"resolver_name":"","ipv4_compat":false}},"cluster_name":"cluster_0","timestamp":"2024-02-13T16:15:21.271Z","health_check_failure_event":{"failure_type":"NETWORK","first_check":false},"locality":{"region":"","zone":"","sub_zone":""}}
{"health_checker_type":"THRIFT","host":{"socket_address":{"protocol":"TCP","address":"127.0.0.1","port_value":40002,"resolver_name":"","ipv4_compat":false}},"cluster_name":"cluster_0","timestamp":"2024-02-13T16:15:26.273Z","health_check_failure_event":{"failure_type":"NETWORK","first_check":false},"locality":{"region":"","zone":"","sub_zone":""}}
{"health_checker_type":"THRIFT","host":{"socket_address":{"protocol":"TCP","address":"127.0.0.1","port_value":40002,"resolver_name":"","ipv4_compat":false}},"cluster_name":"cluster_0","timestamp":"2024-02-13T16:15:31.274Z","health_check_failure_event":{"failure_type":"NETWORK","first_check":false},"locality":{"region":"","zone":"","sub_zone":""}}
```

The same has been already mentioned in #28303, but closed without any real outcome.

Since uneject can happen even at failed health check I doubt this was a planned behaviour for `successful_active_health_check_uneject_host`.
For fix `&& transition == HealthTransition::Changed` should be added to the condition in the callback to 
make sure we uneject the host only at that health check which just cleared the `FAILED_ACTIVE_HC` flag.

Config:
```
...
cluster_manager:
  outlier_detection:
    event_log_path: /tmp/thrift_health.log
static_resources:
  listeners:
  - name: listener_0
    address:
      socket_address:
        protocol: TCP
        address: 127.0.0.1
        port_value: 50051
    filter_chains:
    - filters:
      - name: envoy.filters.network.thrift_proxy
        typed_config:
          "@type": type.googleapis.com/envoy.extensions.filters.network.thrift_proxy.v3.ThriftProxy
          stat_prefix: ingress_service
          protocol: BINARY
          transport: FRAMED
          payload_passthrough: true
          route_config:
            ...
          thrift_filters:
          - name: envoy.filters.thrift.router
            typed_config:
              "@type": type.googleapis.com/envoy.extensions.filters.network.thrift_proxy.router.v3.Router
              close_downstream_on_upstream_error: false
  clusters:
  - name: cluster_0
    ...
    common_lb_config:
      healthy_panic_threshold:
        value: 0.0
    health_checks:
    - timeout: "5s"
      interval: "5s"
      no_traffic_interval: "10s"
      unhealthy_threshold: 3
      healthy_threshold: 1
      event_log_path: /tmp/thrift_health.log
      always_log_health_check_failures: true
      custom_health_check:
        name: envoy.health_checkers.thrift
        typed_config:
          "@type": type.googleapis.com/envoy.extensions.health_checkers.thrift.v3.Thrift
          method_name: ping
          transport: FRAMED
          protocol: BINARY
    outlier_detection:
      consecutive_5xx: 2
      interval: "10s"
      base_ejection_time: "30s"
      max_ejection_time: "30s"
      max_ejection_percent: 100
```

### PR DESCRIPTION
**Pull request by tkovacs-2**
Commit Message:
In sceario described in issue #32442 successful_active_health_check_uneject_host
causes not wanted outlier uneject

Now the health check callbacks get the result of the current hc run (Healthy/Unhealthy)
and based on this callback set in `Upstream::Outlier::DetectorImpl::initialize()`
will uneject really at just successful hc.

Additional Description:
Risk Level: Minor
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
Fixes #32442

### CODE REVIEW

File: source/common/upstream/outlier_detection_impl.cc
Comment (ramaraochavali) [inline source/common/upstream/outlier_detection_impl.cc:315]: This would delay the uneject based on the heath check configuration (interval, healthy threshold etc).  This would essentially render the feature useless - by the time, outlier would have unejected because elapsed time is passed

Comment (tkovacs-2) [reply source/common/upstream/outlier_detection_impl.cc:315]: ??? The feature is useless today: It unejects the host at the next hc run after the eject even if the result was failure but we still didn't reach the unhealthy threshold.
The feature is named: **successful**_active_health_check_uneject_host. For me this means **only** successful hc can uneject.
Please explain why good to uneject for failed hc. I abandon the pr and the issue in that second.

Comment (ramaraochavali) [reply source/common/upstream/outlier_detection_impl.cc:315]: I do not think it will uneject if the health check fails. It should only uneject if atleast one health check succeeds. Are you seeing failed health check unejecting? Atleast the tests I have added when I implemented this ensures the uneject happens on first succesful healthcheck.

Comment (tkovacs-2) [reply source/common/upstream/outlier_detection_impl.cc:315]: Yes, I see failed health check unejecting. See the referenced issue for details.
Right, your test is there and ensures the uneject happens on first succesful healthcheck, but there is nothing about failed healthcheck. It is not even possible make such today since callback doesn't get this information.
If unhealthy_threshold is 3 let's say, this hc sequence can happen:

success, !FAILED_ACTIVE_HC
fail, !FAILED_ACTIVE_HC
fail, !FAILED_ACTIVE_HC
fail, FAILED_ACTIVE_HC
fail, FAILED_ACTIVE_HC
...

Callback just checks !FAILED_ACTIVE_HC. If the ejection happens after the 1st or 2nd hc run then next is fail but will uneject.
FAILED_ACTIVE_HC is the host's health flag and in general it is not equal with the current hc run result. Do I misunderstand something?

I tested this with 1.28.0, but I can try double check with the latest. Also I'm using thrift health check but I think that shouldn't matter.

Comment (ramaraochavali) [reply source/common/upstream/outlier_detection_impl.cc:315]: But if you check `healthy && **!host->healthFlagGet(Host::HealthFlag::FAILED_ACTIVE_HC**` then you will only uneject on host transitioning to healthy via active health checks(which might be 3 or 10). - That was my point. But active health checks are configured in such a way to the threshold will finish before ejection time - it may fine otherwise this feature wont be of that value.,

Comment (ramaraochavali) [reply source/common/upstream/outlier_detection_impl.cc:315]: Yes.  #3 was the original intention. Now that I think of it more, it is OK to honour health check configuration and uneject only if FAILED_ACTIVE_HC is not set to align with host health check flag. But we need to document that when this feature is used, it will only be useful if health check is configured in a way that total time for successful health check is less than the ejection time.

Hope this makes sense. @wbpcode  @zuercher  WDYT?

**Review by zuercher — COMMENTED** [review]
I think this change is fine, but would like to see the behavior documented somewhere. For example in `docs/root/intro/arch_overview/upstream/outlier.rst`.

File: docs/root/intro/arch_overview/upstream/outlier.rst
Comment (zuercher) [inline docs/root/intro/arch_overview/upstream/outlier.rst:209]: If you'll allow me, I think this reads a little better:

```suggestion
  clears all outlier detection counters. If the host is failing health checks, but has not reached the
  :ref:`unhealthy_threshold<envoy_v3_api_field_config.core.v3.HealthCheck.unhealthy_threshold>`, a single successful health check
  will uneject the host. If the FAILED_ACTIVE_HC health flag is set for the host,
  :ref:`healthy_threshold<envoy_v3_api_field_config.core.v3.HealthCheck.healthy_threshold>` consecutive successful health checks
  will uneject the host (and clear the FAILED_ACTIVE_HC flag).
  If your active health check is not validating data plane traffic then in situations where
```

Comment (tkovacs-2) [reply docs/root/intro/arch_overview/upstream/outlier.rst:209]: The first is part little confusing in this way.  Previous failed checks are not condition.
How about: 'If the host has not reached ... failing health checks, a single succesful ...' ?

Comment (zuercher) [reply docs/root/intro/arch_overview/upstream/outlier.rst]: Sure.

**Review by zuercher — APPROVED** [review]

**Comment by soulxu** [comment]
@zuercher is it ready to merge or you are waiting for someone to take a look at this more?

**Comment by zuercher** [comment]
Needs a second maintainer to approve and then it can be merged.

@ramaraochavali?

**Comment by RyanTheOptimist** [comment]
@zuercher as per the new policy, I don't think we need a second review on this:
https://github.com/envoyproxy/envoy/pull/30995/files
What do you think?

**Comment by zuercher** [comment]
Sorry, missed that change. Yes we can merge, though I realized there should be a release note for this bug fix.

**Comment by RyanTheOptimist** [comment]
/wait
For updated release notes

**Review by zuercher — APPROVED** [review]

File: changelogs/current.yaml
Comment (phlax) [inline changelogs/current.yaml:243]:
```suggestion
    :ref:`successful_active_health_check_uneject_host <envoy_v3_api_field_config.cluster.v3.OutlierDetection.successful_active_health_check_uneject_host>`.
```

File: changelogs/current.yaml
Comment (phlax) [inline changelogs/current.yaml:244]:
```suggestion
    Before, if ``FAILED_ACTIVE_HC`` health flag was set, a failed health check could uneject the host.
```

**Review by phlax — COMMENTED** [review]
@tkovacs-2 a couple of changelog nits and this should be good to land

File: changelogs/current.yaml
Comment (phlax) [reply changelogs/current.yaml:244]: im still a bit confused as to meaning wrt what is set/happens

is this correct?

>  Before, a failed health check could uneject the host if the `FAILED_ACTIVE_HC` had not been set

?

Comment (tkovacs-2) [reply changelogs/current.yaml]: Ok, I just added back the `health flag` phrase. I think it is useful.

File: changelogs/current.yaml
Comment (tkovacs-2) [reply changelogs/current.yaml]: I don't get the intention of this space.
I have changed the line breaks, please check if ok in this way.

**Review by phlax — DISMISSED** [review]
lgtm, thanks @tkovacs-2

**Review by phlax — APPROVED** [review]

**Pull request merged**
